src/builder/List.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IQuarkWallet} from "quark-core/src/interfaces/IQuarkWallet.sol";
+import {Actions} from "./Actions.sol";
+
+library List {
+    error IndexOutOfBound();
+
+    struct ListStuct {
+        bytes[] bytesArray;
+        uint256 size;
+    }
+
+    function newList() internal pure returns (ListStuct memory) {
+        return ListStuct(new bytes[](0), 0);
+    }
+
+    function toArray(ListStuct memory list) internal pure returns (bytes[] memory) {
+        bytes[] memory result = new bytes[](list.size);
+        for (uint256 i = 0; i < list.size; ++i) {
+            result[i] = list.bytesArray[i];
+        }
+        return result;
+    }
+
+    function add(ListStuct memory list, bytes memory item) internal pure {
+        if ((list.size + 1) * 2 > list.bytesArray.length) {
+            bytes[] memory newBytesArray = new bytes[]((list.bytesArray.length + 1) * 2);
+            for (uint256 i = 0; i < list.size; ++i) {
+                newBytesArray[i] = list.bytesArray[i];
+            }
+            newBytesArray[list.size] = item;
+
+            list.bytesArray = newBytesArray;
+            list.size = list.size + 1;
+        } else {
+            list.bytesArray[list.size] = item;
+            list.size = list.size + 1;
+        }
```

Comment:
This can be simplified to something like:

```
  function addItem(DynamicArray memory array, bytes memory item) internal {
        if (array.items.length == array.length) {
            bytes[] memory newItems = new bytes[](array.length * 2 + 1);
            for (uint256 i = 0; i < array.length; ++i) {
                newItems[i] = array.items[i];
            }
            array.items = newItems;
        }
        array.items[array.length] = item;
        array.length++;
  }
```

Reply:
thanks for the nice simplified version of the logics haha.